flink-table-store-flink/flink-table-store-flink-common/src/main/java/org/apache/flink/table/store/connector/action/MergeIntoAction.java

```diff
@@ -0,0 +1,678 @@
+                        + "based on merge-condition and optional not-matched-by-source-condition.");
+        System.out.println(
+                "  6. If both matched-upsert and matched-delete actions are present, their conditions must both be present too "
+                        + "(same to not-matched-by-source-upsert and not-matched-by-source-delete). Otherwise, all conditions are optional.");
+        System.out.println("  7. source-alias cannot be duplicated with existed table name.");
+        System.out.println();
+
+        System.out.println("Examples:");
+        System.out.println(
+                "  merge-into --path hdfs:///path/to/T\n"
+                        + "             --using-table S\n"
+                        + "             --on \"T.k=S.k\"\n"
+                        + "             --merge-actions matched-upsert\n"
+                        + "             --matched-upsert-condition \"T.v<>S.v\"\n"
+                        + "             --matched-upsert-set \"S.k,S.v\"");
+        System.out.println(
+                "  It will find matched rows of target table that meet condition (T.k=S.k AND T.v<>S.v) ind '-D' to test_table, then update T.v with S.v.");
+    }
+
+    @Override
+    public void run() throws Exception {
+        // prepare source table
+        if (sourceTable != null) {
+            if (sourceTable.contains(".")) {
+                sourceTableIdentifier = Identifier.fromString(sourceTable);
+            } else {
+                sourceTableIdentifier =
+                        Identifier.create(identifier.getDatabaseName(), sourceTable);
+            }
+        } else {
+            tEnv.registerTable(sourceAlias, tEnv.sqlQuery(source));
+            sourceTableIdentifier = Identifier.create(identifier.getDatabaseName(), sourceAlias);
+        }
+
+        List<DataStream<RowData>> dataStreams = new ArrayList<>();
+        if (matchedUpsert || notMatchedUpsert) {
+            dataStreams.add(getUpsertDataStream());
+        }
+
+        if (matchedDelete || notMatchedDelete) {
+            dataStreams.add(getDeleteDataStream());
+        }
+
+        if (insert) {
+            dataStreams.add(getInsertDataStream());
+        }
+
+        DataStream<RowData> unionDs = dataStreams.get(0);
+        sink(unionDs.union(dataStreams.stream().skip(1).toArray(DataStream[]::new)));
+    }
+
+    private DataStream<RowData> getUpsertDataStream() {
+        Table updateSource;
+        if (matchedUpsert && notMatchedUpsert) {
+            updateSource = getMatchedUpsertSource().unionAll(getNotMatchedUpsertSource());
+        } else if (matchedUpsert) {
+            updateSource = getMatchedUpsertSource();
+        } else {
+            updateSource = getNotMatchedUpsertSource();
+        }
+
+        return toDataStream(updateSource, RowKind.UPDATE_AFTER, converters);
+    }
+
+    private Table getMatchedUpsertSource() {
```

Comment:
Can we just return a `Optional<Table>`.
```
if (!matchedUpsert) {
   return Optional.empty;
}
Table table = ...;
return Optional.of(table);
```

matchedUpserts can be hide in own methods.